### DISCUSSION TITLE
MolOps::getMolFrags returns std::vector<boost::shared_ptr<ROMol>>

### DISCUSSION THREAD

**Discussion by DavidACosgrove** [rdkit/rdkit#7822] [Development]
Hi,

I am trying to wean myself off boost smart pointers and onto std smart pointers using std::<unique_ptr> for preference.  Currently I am fragmenting molecules, and RDKit::MolOps::getMolFrags() returns std::vector<boost::shared_ptr<ROMol>>.  This leaves me with a nasty mixture in the code.  Are there any plans to change this?  Will there be a v2:: api, for example?  In fact, since there are only 2 functions in MolOps that return shared_ptr, can I make a v2:: api that returns std::vector<std::unique_ptr<ROMol>> instead?

Dave

**Comment by e-kwsm** [comment]
unfortunately boost::serialization does not support std::shared_ptr
#6923

**Reply by bp-kelley** [reply]
@e-kwsm serializing std::shared_ptr is a solvable problem it just isn't support directly by boost and would require some wrapping to ensure that the old and new formats are compatible.  It would be a major API change though.

@DavidACosgrove the reason that boost::shared_ptr is prevalent where it would not be the current first choice is because unique_ptr simply didn't exist at the time and shared_ptr was introduced in boost itself. 

I think we can make the case for a v2 and unique_ptrs, but in honesty, vectors of shared_ptrs integrate with python a lot better.  What happens when you copy a bunch of unique ptrs for instance?   

That being said, if I were designing the API now, it would be std::vector<unique_ptr<T>> in C++ that get converted to Python controlled raw or shared ptrs in the wrapper.
